r/src/json.cpp

```diff
@@ -31,14 +31,24 @@ std::shared_ptr<arrow::json::ReadOptions> json___ReadOptions__initialize(bool us
 }
 
 // [[arrow::export]]
-std::shared_ptr<arrow::json::ParseOptions> json___ParseOptions__initialize(
+std::shared_ptr<arrow::json::ParseOptions> json___ParseOptions__initialize1(
     bool newlines_in_values) {
   auto res =
       std::make_shared<arrow::json::ParseOptions>(arrow::json::ParseOptions::Defaults());
   res->newlines_in_values = newlines_in_values;
   return res;
 }
 
+// [[arrow::export]]
+std::shared_ptr<arrow::json::ParseOptions> json___ParseOptions__initialize2(
+    bool newlines_in_values, const std::shared_ptr<arrow::Schema>& explicit_schema) {
+  auto res =
+    std::make_shared<arrow::json::ParseOptions>(arrow::json::ParseOptions::Defaults());
+  res->newlines_in_values = newlines_in_values;
+  res->explicit_schema = explicit_schema;
+  return res;
+}
+
```

Comment:
Possibly silly question because I'm a C++ noob:
Would it work to keep this as just one function `json___ParseOptions__initialize` with `explicit_schema` defaulting to `nullptr` and then do
```cpp
if (explicit_schema != nullptr) {
  res->explicit_schema = explicit_schema;
}
```

Reply:
I gave this a try, but when I call `JsonParseOptions$create`, it results in the following error:
` Error: Invalid R object for std::shared_ptr<arrow::Schema>, must be an ArrowObject` 

I am guessing it's because the code below passes in `NULL` as the schema argument, and even with `explicit_schema` defaulting to `nullptr`, that default value is never used?
```
JsonParseOptions$create <- function(newlines_in_values = FALSE, schema = NULL) {
    json___ParseOptions__initialize(newlines_in_values, schema)
}
```

Reply:
Cool, thanks. I'm confused about why this seems to work in a couple of the `dataset___FileSystemDatasetFactory__Make*` functions in `dataset.cpp` but it doesn't work here.

Reply:
It might be the way I've implemented it; I'll check there and see how they compare.

Reply:
R's `NULL` is not the same as `nullptr`. It works in dataset___FileSystemDatasetFactory__Make2 because Make1 calls it with nullptr. (The nullptr check you just added in Make3 will never be nullptr (nor `NULL`) because it only is called from R if the input is a `Partitioning` object.

Reply:
Gotcha. FTR, I did not add that `nullptr` check in `Make3` 🙂

Reply:
My bad, I thought I saw that you copied it from the other function (because I noticed the `TODO(fsaintjacques)` duplicated in the diff) but maybe it was just next to what you were doing

Reply:
Useful stuff to me to learn, thanks for flagging this up and having this discussion!  Just to check where this is, off the back of the discussion above, it is fine to leave it as it is currently, or shall I make some more changes?

Reply:
It's good.